### ISSUE TITLE
v3: Handling of keys (strings, symbols, both)

### ISSUE THREAD

**Issue by RST-J** [ruby-json-schema/json-schema#243]
To pick up on the discussion about hash keys that started in #161: How do we want to handle hash keys in the next major release?

I think, this should be changed soon because its a very central issue that affects nearly everything to come.

I suggest, that we internally always use string keys in every data structure (data and schemas).
If we decide to drop support for symbols, then this is it. But this will raise problems with insert defaults (#184) if we decide to keep support for that.

Otherwise we need to provide an option that indicates whether to convert all keys in the data to symbols after the validation process if it succeeded.
I'm in favour of the latter (if there are no big obstacles along the way), because its easy for the user and simplifies insert defaults (if still supported).

Quotes from @iainbeeston from #161
> 1. If we symbolize keys, we should also symbolize any other reference to properties in the schema (for example, required should have a symbol or array of symbols, so should depends).

My idea is to have a flag ``:symbolize_keys`` which defaults to ```false``` and if set to ```true``` converts the keys in the validated data to symbols **after** validation. During validation everything will be strings, so we do not need to convert anything in schemas. If we do not and use cached schemas, we'd have to convert the schemas used all the time or store them in a symbolized and a stringified version.

Thoughts?

**Comment by iainbeeston** [comment]
I think having a :symbolize_keys option is a good idea.

If using strings internally causes problems for defaults I would prefer dropping support for defaults.

There is another option, which is that we don't convert symbols to strings at all. JSON has no concept of strings, and no Json parser that I'm aware of produces symbol keys, so I suspect that symbolized hashes are rarely used in practice. It's also easy (and less error probe) for a caller to stringify keys than it is for us to do it automatically. So what I'm suggesting is that we don't do the conversion at all and raise an error if we find a symbol key. This would be a breaking change, but then this is for the v3 release, right?

**Comment by RST-J** [comment]
Yes. I actually think, this should be for v3 no matter what we decide to do because this will likely break current behaviour in some way.

I can also imagine that it would be easier to remove support for symbols (and also everything related to defaults) first, even if we want to have it back in in some way later, because the core validation code gets cleaned up.

**Comment by RST-J** [comment]
Ok, after looking into this it seems that it isn't as huge as I expected it to be and more important we already used [string keys](https://github.com/ruby-json-schema/json-schema/blob/3.x/lib/json-schema/validator.rb#L580) during validation.

Actually, the tests will still pass with only the changes in the attribute classes.

What remains mixed is [inserting defaults](https://github.com/ruby-json-schema/json-schema/blob/3.x/lib/json-schema/validator.rb#L433) where symbol keys in the data are honoured.

So with this new insight, what do you think?

**Comment by iainbeeston** [comment]
I think this is one of the issues with inserting defaults... I'd rather have the code simplified by removing that feature, to be honest. But if we're going to have them then your suggestion (for giving the user an option to control what type is used for hash keys) is a good idea.

**Comment by iainbeeston** [comment]
By the way, I think there is still the issue of properties that refer to other properties, like depends and required. By changing keys from symbols to strings I think these referential properties might no longer point to their original target. It's possible that we should turn all symbols into strings throughout the schema (again, Json doesn't have a concept of symbols).
